File: algorithms/396.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxRotateFunction(vector<int> &A) {
        int s = 0, t = 0;
        for (int i = 0; i < A.size(); i++) {
            s += A[i];
            t += i*A[i];
        }
        int result = t;
        for (int i = A.size()-1; i >= 1; i--) {
            t = t + s - A[i]*A.size();
            result = max(result, t);
        }
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/python/396.py
class Solution:
    def maxRotateFunction(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        l, s = len(A), sum(A)
        if l == 0:
            return 0
        f = [0] * l
        f[0] = sum([i*A[i] for i in range(l)])
        for i in range(1, l):
            f[i] = f[i-1] + s - l * A[l-i]
        return max(f)
